draft-ietf-tls-esni.md

```diff
@@ -413,10 +413,10 @@ key_share
 used to derive the ESNI key.
 
 record_digest
-: A cryptographic hash of the ESNIKeys structure from which the ESNI
-key was obtained, i.e., from the first byte of "checksum" to the end
-of the structure.  This hash is computed using the hash function
-associated with `suite`.
+: A cryptographic hash of the ESNIKeys structure, excluding the "extensions"
+field, from which the ESNI key was obtained, i.e., from the first byte of
+"checksum" to the last byte of "not_after". This hash is computed using the
+hash function associated with `suite`.
```

Comment:
It seems weird to exclude other extensions. Also I assume you mean the first byte of "version"?

Reply:
I agree. I'm not sure this is the right approach. It seems we should either (a) have individual extensions included in the digest or not via some mandatory-like bit, or (b) have a separate extensions field that's omitted from the digest, in which the address information can go. Both are... not great.

Reply:
To me the point of `record_digest` is to identify a specific key so that the TLS server doesn't have to do trial decryption, that is, its point is not the digest itself, but the fact that it identifies a key, is this correct?

If so, it seems fine (but not great) that it would only really cover the specific key and the metadata directly associated with the key, while extensions can potentially be associated with multiple keys.

But a potentially better alternative is to just get rid of `record_digest` (if my assumption that it only provides an "identity" for the key), and replace with an explicit `key_identity` field that is generated by the server and would get added to `ESNIKeys` as well as replace `record_digest` in the ClientHello extension.

Thoughts?

Reply:
@ghedo 
> But a potentially better alternative is to just get rid of `record_digest` (if my assumption that it only provides an "identity" for the key), and replace with an explicit `key_identity` field that is generated by the server and would get added to `ESNIKeys` as well as replace `record_digest` in the ClientHello extension.

I am not sure if that is the correct thing to do. I'd be concerned that omission of a digest might lead to availability concerns.

The reason we have had `record_digest` is because the TLS handshake used to always hard-fail when client uses a corrupt ESNI record. With #124 being merged, we have a fallback mechanism that would work in _some_ failure modes without the need for the digest. However, I believe that there are still certain cases of hard failure (like the server sending an `illegal_parameter` alert) when the ESNI record is corrupt.

Reply:
I'm not sure I follow @kazuho, isn't the `checksum` value intended to detect record corruption? If so, why do we have two?

Reply:
@ghedo `record_digest` was originally added to prevent downgrades, so we want it to cover as much of the necessary bits in the record as possible. It just so happens to also serve as a key identifier for the server.